File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarBackupManager.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

private let TAG = "PolarDeviceBackup"
private let ARABICA_SYS_FOLDER = "/SYS/"
private let ARABICA_USER_ROOT_FOLDER = "/U/0/"
private let USER_WILD_CARD_ROOT_FOLDER = "/U/*/"
private let WILD_CARD_CHARACTER = "*"

public class PolarBackupManager {

    public struct BackupFileData {
        let data: Data
        let directory: String
        let fileName: String
    }

    private struct DeviceFolderEntry {
        let name: String
        let size: Int64
        let folderPath: String
    }

    let client: BlePsFtpClient

    public init(client: BlePsFtpClient) {
        self.client = client
    }

    public func backupDevice() async throws -> [BackupFileData] {
        BleLogger.trace("backupDevice() called")
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = ARABICA_SYS_FOLDER }
        do {
            let content = try await client.request(try operation.serializedBytes())
            let parentDirEntries = try Protocol_PbPFtpDirectory(serializedBytes: content as Data)
            let entries = parentDirEntries.entries.map { ARABICA_SYS_FOLDER + $0.name }
            var backupDirectories: [String] = []
            if let backupEntry = entries.first(where: { $0.hasSuffix("BACKUP.TXT") }) {
                BleLogger.trace("Found BACKUP.TXT: \(backupEntry)")
                let backupData = try await loadFile(path: backupEntry)
                let data = Data(backupData)
                let stream = InputStream(data: data)
                var accumulatedString = ""
                stream.open()
                defer { stream.close() }
                let bufferSize = 1024
                let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
                defer { buffer.deallocate() }
                while stream.hasBytesAvailable {
                    let bytesRead = stream.read(buffer, maxLength: bufferSize)
                    if bytesRead > 0, let chunk = String(bytesNoCopy: buffer, length: bytesRead, encoding: .utf8, freeWhenDone: false) {
                        accumulatedString += chunk
                        while let range = accumulatedString.range(of: "\n") {
                            let line = String(accumulatedString[..<range.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
                            backupDirectories.append(line)
                            accumulatedString.removeSubrange(..<range.upperBound)
                        }
                    }
                }
            } else {
                BleLogger.error("No BACKUP.TXT found, using default backup directories")
            }
            addDefaultBackupDirectories(to: &backupDirectories)
            BleLogger.trace("Backup directories found: \(backupDirectories)")
            var result = [BackupFileData]()
            for dir in backupDirectories {
                BleLogger.trace("Processing backup directory: \(dir)")
                let files = try await backupDirectory(backupDirectory: dir)
                BleLogger.trace("Backup directory ready: \(dir)")
                result.append(contentsOf: files)
            }
            return result
        } catch {
            BleLogger.error("Failed to get backup content, error: \(error)")
            return []
        }
    }

    private func addDefaultBackupDirectories(to directories: inout [String]) {
        let defaults = ["/U/*/S/PHYSDATA.BPB", "/U/*/S/UDEVSET.BPB", "/U/*/S/PREFS.BPB", "/U/*/USERID.BPB"]
        for dir in defaults { if !directories.contains(dir) { directories.append(dir) } }
    }

    public func restoreBackup(backupFiles: [BackupFileData]) async throws {
        BleLogger.trace("Starting backup restoration for \(backupFiles.count) files")
        var failedFiles: [(fileName: String, error: Error)] = []
        for backupFileData in backupFiles {
            BleLogger.trace("Restoring: \(backupFileData.fileName)")
            do {
                var operation = Protocol_PbPFtpOperation()
                operation.command = .put
                operation.path = backupFileData.directory + backupFileData.fileName
                let header = try operation.serializedData() as NSData
                let dataStream = InputStream(data: backupFileData.data)
                for try await bytes in client.write(header, data: dataStream) {
                    BleLogger.trace("Writing firmware update file: \(backupFileData.directory)\(backupFileData.fileName), bytes to write: \(bytes)")
                }
                BleLogger.trace("Restored: \(backupFileData.directory)\(backupFileData.fileName)")
            } catch {
                BleLogger.error("Failed to restore \(backupFileData.directory)\(backupFileData.fileName): \(error)")
                failedFiles.append((fileName: backupFileData.directory + backupFileData.fileName, error: error))
            }
        }
        if !failedFiles.isEmpty {
            let summary = failedFiles.map {"\($0.fileName): \($0.error)" }.joined(separator: ", ")
            BleLogger.error("Backup restoration completed with \(failedFiles.count) failure(s): \(summary)")
            throw NSError(
                domain: "PolarBackupManager",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Restore failed for \(failedFiles.count) file(s): \(summary)"]
            )
        }
        BleLogger.trace("Backup restoration complete")
    }

    private func loadFile(path: String) async throws -> [UInt8] {
        var operation = Protocol_PbPFtpOperation()
        operation.command = .get
        operation.path = path
        let data = try await client.request(try operation.serializedBytes())
        return [UInt8](data)
    }

    private func backupDirectory(backupDirectory: String) async throws -> [BackupFileData] {
        let directoryPaths = backupDirectory.components(separatedBy: .newlines)
        var result = [BackupFileData]()
        for directoryPath in directoryPaths {
            guard !directoryPath.isEmpty else { continue }
            let path = directoryPath.replacingOccurrences(of: USER_WILD_CARD_ROOT_FOLDER, with: ARABICA_USER_ROOT_FOLDER)
            if path.contains(WILD_CARD_CHARACTER) {
                let rootPath = path.components(separatedBy: WILD_CARD_CHARACTER).first ?? ""
                let entries = try await fetchRecursively(path: rootPath)
                let subdirs = entries.filter { $0.folderPath.isFolder }.map { rootPath + $0.name }
                for subdir in subdirs {
                    let files = try await backUpDirectories([subdir])
                    result.append(contentsOf: files)
                }
            } else if path.isFolder {
                let subDirs = try await loadSubDirectories(path: path)
                let files = try await backUpDirectories(subDirs)
                result.append(contentsOf: files)
            } else {
                do {
                    let data = try await loadFile(path: path)
                    let file = (path as NSString).lastPathComponent
                    let filePath = (path as NSString).deletingLastPathComponent + "/"
                    result.append(BackupFileData(data: Data(data), directory: filePath, fileName: file))
                } catch {
                    BleLogger.error("Error loading file: \(path), error: \(error)")
                }
            }
        }
        return result
    }

    private func fetchRecursively(path: String) async throws -> [DeviceFolderEntry] {
        var operation = Protocol_PbPFtpOperation()
        operation.command = .get
        operation.path = path
        let request = try operation.serializedData()
        var entries = [DeviceFolderEntry]()
        do {
            let data = try await client.request(request)
            let dir = try Protocol_PbPFtpDirectory(serializedBytes: data as Data)
            for entry in dir.entries {
                entries.append(DeviceFolderEntry(name: (entry.name as NSString).lastPathComponent, size: Int64(entry.size), folderPath: entry.name))
                if entry.name.hasSuffix("/") {
                    let subEntries = try await fetchRecursively(path: path + entry.name)
                    entries.append(contentsOf: subEntries)
                }
            }
        } catch {
            BleLogger.error("fetchRecursively() error: \(error)")
        }
        return entries
    }

    private func loadSubDirectories(path: String) async throws -> [String] {
        return try await fetchRecursively(path: path).map { path + $0.name }
    }

    private func backUpDirectories(_ folders: [String]) async throws -> [BackupFileData] {
        var result = [BackupFileData]()
        for folder in folders {
            let files = try await backupDirectory(backupDirectory: folder)
            result.append(contentsOf: files)
        }
        return result
    }
}

extension String {
    var isFolder: Bool { return hasSuffix("/") }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarBackupManager.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import java.io.ByteArrayInputStream
import java.io.File
import protocol.PftpRequest
import protocol.PftpResponse.PbPFtpDirectory

private const val TAG = "PolarDeviceBackup"
private const val ARABICA_SYS_FOLDER = "/SYS/"
private const val ARABICA_USER_ROOT_FOLDER = "/U/0/"
private const val USER_WILD_CARD_ROOT_FOLDER = "/U/*/"
private const val WILD_CARD_CHARACTER = "*"

/**
 * Handles backing up the device.
 * @param client BlePsFtpClient
 */
class PolarBackupManager(private val client: BlePsFtpClient) {

    class BackupFileData(
        val data: ByteArray,
        val directory: String,
        val fileName: String
    )

    private class DeviceFolderEntries(val entriesList: List<DeviceFolderEntry>)

    private data class DeviceFolderEntry(
        val name: String,
        val size: Long = 0L,
        val folderPath: String = ""
    )

    /**
     * Backs up the device.
     * @return List of backed up file data.
     */
    suspend fun backupDevice(): List<BackupFileData> {
        BleLogger.d(TAG, "Backing up device")
        BleLogger.d(TAG, "Requesting backup content")

        val builder = PftpRequest.PbPFtpOperation.newBuilder()
            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
            .setPath(ARABICA_SYS_FOLDER)

        val response = client.request(builder.build().toByteArray())
        BleLogger.d(TAG, "Received response from client request")
        val dir = PbPFtpDirectory.parseFrom(response.toByteArray())
        val entries = dir.entriesList.map { entry ->
            BleLogger.d(TAG, "Entry added: ${entry.name}, size: ${entry.size}")
            ARABICA_SYS_FOLDER + entry.name to entry.size
        }

        BleLogger.d(TAG, "Entries retrieved: ${entries.size}")

        val defaultBackupDirectories = linkedSetOf(
            "/U/*/S/PHYSDATA.BPB",
            "/U/*/S/UDEVSET.BPB",
            "/U/*/S/PREFS.BPB",
            "/U/*/USERID.BPB"
        )

        val backupEntry = entries.find { it.first.endsWith("BACKUP.TXT") }

        val lines: MutableSet<String> = if (backupEntry != null) {
            BleLogger.d(TAG, "Backup Entry: ${backupEntry.first}, Size: ${backupEntry.second}")
            val data = loadFile(backupEntry.first)
            BleLogger.d(TAG, "Data loaded for backup entry, size: ${data.size}")
            val stream = ByteArrayInputStream(data)
            val reader = stream.bufferedReader()
            val readLines = reader.useLines { seq ->
                seq.mapNotNull { line ->
                    try {
                        BleLogger.d(TAG, "Reading line: $line")
                        line
                    } catch (e: Exception) {
                        BleLogger.e(TAG, "Failed to read line: $line, error: $e")
                        null
                    }
                }.toMutableSet()
            }
            readLines.addAll(defaultBackupDirectories)
            BleLogger.d(TAG, "Lines read from backup entry plus defaults: ${readLines.size}")
            readLines
        } else {
            BleLogger.w(TAG, "Device does not have BACKUP.TXT, using default backup directories")
            defaultBackupDirectories
        }

        return try {
            lines.flatMap { line ->
                BleLogger.d(TAG, "Backing up line: $line")
                try {
                    backupDirectory(line)
                } catch (error: Throwable) {
                    BleLogger.e(TAG, "Failed to backup line: $line, error: $error")
                    emptyList()
                }
            }
        } catch (it: Throwable) {
            BleLogger.e(TAG, "Failed to get backup content, error: $it")
            emptyList()
        }
    }

    /**
     * Restores the backup files.
     * @param backupFiles List of BackupFileData to be restored.
     */
    suspend fun restoreBackup(backupFiles: List<BackupFileData>) {
        backupFiles.forEach {
            BleLogger.d(TAG, "Restoring backup file: ${it.directory} + ${it.fileName}")
        }
        for (backupFileData in backupFiles) {
            val header = PftpRequest.PbPFtpOperation.newBuilder()
                .setCommand(PftpRequest.PbPFtpOperation.Command.PUT)
                .setPath(backupFileData.directory + backupFileData.fileName)
                .build().toByteArray()
            val dataStream = ByteArrayInputStream(backupFileData.data)
            BleLogger.d(TAG, "Sending PftpRequest: ${header.toString(Charsets.UTF_8)}")
            try {
                client.write(header, dataStream).collect { }
                BleLogger.d(TAG, "Successfully restored backup file: ${backupFileData.fileName}")
            } catch (error: Throwable) {
                BleLogger.e(TAG, "Failed to restore backup file: ${backupFileData.fileName}, error: $error")
                // continue with next file
            }
        }
    }

    private suspend fun loadFile(path: String): ByteArray {
        return try {
            client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(path)
                    .build().toByteArray()
            ).toByteArray()
        } catch (error: Throwable) {
            BleLogger.i(TAG, "Failed to load file: $path, error: $error")
            throw error
        }
    }

    private suspend fun backupDirectory(backupDirectory: String): List<BackupFileData> {
        val path = backupDirectory.replace(USER_WILD_CARD_ROOT_FOLDER, ARABICA_USER_ROOT_FOLDER)

        return when {
            path.contains(WILD_CARD_CHARACTER) -> {
                val rootPath = path.substringBefore(WILD_CARD_CHARACTER)
                val subFolderToBackup = path.substringAfter(WILD_CARD_CHARACTER)
                    .removePrefix(File.separator)
                    .split(File.separator)
                    .firstOrNull()

                val backupDirectories = try {
                    loadSubDirectories(rootPath)
                        .filter { it.isFolder() }
                        .flatMap { folder ->
                            loadEntries(folder).entriesList.filter {
                                val name = it.name.replace(File.separator, "")
                                name == subFolderToBackup
                            }.map { folder + it.name }
                        }
                } catch (e: Exception) {
                    BleLogger.e(TAG, "Error loading subdirectories: $rootPath, error: $e")
                    emptyList()
                }

                backUpDirectories(backupDirectories)
            }

            path.isFolder() -> {
                try {
                    backUpDirectories(loadSubDirectories(path))
                } catch (e: Exception) {
                    BleLogger.e(TAG, "Error loading subdirectories: $path, error: $e")
                    emptyList()
                }
            }

            else -> {
                BleLogger.d(TAG, "Backup file: $path")
                try {
                    val data = loadFile(path)
                    val file = path.substringAfterLast(File.separator)
                    val filePath = path.substringBefore(file)
                    listOf(BackupFileData(data, filePath, file))
                } catch (it: Throwable) {
                    BleLogger.e(TAG, "Error loading file: $path, error: $it")
                    emptyList()
                }
            }
        }
    }

    private suspend fun loadEntries(path: String): DeviceFolderEntries {
        val entriesList = mutableListOf<DeviceFolderEntry>()
        fetchRecursively(path).forEach { (entryPath, size) ->
            val name = entryPath.substringAfterLast('/')
            entriesList.add(DeviceFolderEntry(name, size, entryPath))
        }
        return DeviceFolderEntries(entriesList)
    }

    private suspend fun fetchRecursively(path: String): List<Pair<String, Long>> {
        val builder = PftpRequest.PbPFtpOperation.newBuilder()
            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
            .setPath(path)
        val response = client.request(builder.build().toByteArray())
        val dir = PbPFtpDirectory.parseFrom(response.toByteArray())
        val entries = dir.entriesList.associate { path + it.name to it.size }

        if (entries.isEmpty()) return emptyList()

        return entries.flatMap { (entryPath, size) ->
            if (entryPath.endsWith("/")) {
                fetchRecursively(entryPath)
            } else {
                listOf(entryPath to size)
            }
        }
    }

    private suspend fun loadSubDirectories(path: String): List<String> {
        return loadEntries(path).entriesList.map { path + it.name }
    }

    private suspend fun backUpDirectories(folders: List<String>): List<BackupFileData> {
        return folders.flatMap { path -> backupDirectory(path) }
    }

    private fun String.isFolder() = this.endsWith("/")
}